feat(integrations): create digiwf-message lib to replace spring-cloud-stream-utils lib and simplify integration development

lmoesle opened the pull request:
**Description**

- Moved spring-cloud-stream-utils to digiwf-message lib
- Created a digiwf-integrations-lib to use custom annotation
- Implemented custom scs routing with custom annotations
- Created examples and documentation

**Reference**

Issues #167

lmoesle committed:
feat: refactor spring cloud stream utils to be less binder dependent

lmoesle committed:
feat: refactor spring cloud stream utils to be less binder dependent

lmoesle committed:
feat: refactor spring cloud stream utils to be less binder dependent

lmoesle committed:
feat: deprecated services in spring cloud stream utils

lmoesle committed:
feat: added simple example with custom adapter

lmoesle committed:
feat: added simple example for messaging lib

lmoesle committed:
feat: added spring cloud stream example for messaging lib

lmoesle committed:
feat: use proper spring cloud stream

lmoesle committed:
feat: use @HandleIncident and @HandleTechnicalError to handle errors

lmoesle committed:
feat: use @HandleIncident and @HandleTechnicalError to handle errors

lmoesle committed:
feat: removed aspects and spring aop

lmoesle committed:
Merge branch 'dev' into feat/167-create-handle-error-use-case

lmoesle committed:
feat: moved spring cloud stream features to a dedicated adapter module

lmoesle committed:
feat: implemented custom error handler to create incidents and technical errors

lmoesle committed:
Merge branch 'dev' into feat/167-create-handle-error-use-case

lmoesle committed:
feat: added some javadoc

lmoesle committed:
feat: create integration framework

lmoesle committed:
feat: refactor lib structure

lmoesle committed:
feat: refactor lib structure

lmoesle committed:
feat: refactor lib structure

lmoesle committed:
Merge branch 'dev' into feat/167-create-handle-error-use-case

lmoesle committed:
docs: added javadoc and minor refactoring (renaming variables)

lmoesle committed:
test: added Tests for MessageApiImpl and package structure refactoring

lmoesle committed:
test: added tests for process messaging

lmoesle committed:
Merge branch 'dev' into feat/167-create-handle-error-use-case

lmoesle committed:
test: added tests digiwf-integration-lib

lmoesle committed:
test: added digiwf-integrations and digiwf-message lib to coverage module

lmoesle committed:
test: added a few more tests for better coverage

lmoesle commented:
@peterhnm I added the modeler to the release pipeline

lmoesle committed:
feat: moved the example integration to the digiwf-integrations package

lmoesle committed:
docs: how to create an integration

lmoesle committed:
docs: added docs for digiwf-integration-lib

lmoesle committed:
docs: added docs for digiwf-message

lmoesle committed:
Merge branch 'dev' into feat/167-create-handle-error-use-case

lmoesle committed:
feat: validation exceptions are treated as technical errors

lmoesle committed:
feat: improve error handling

lmoesle committed:
Merge branch 'dev' into feat/167-create-handle-error-use-case

lmoesle committed:
Merge branch 'dev' into feat/167-create-handle-error-use-case

lmoesle committed:
chore: version bump

lmoesle committed:
chore: adjusted CHANGELOG.md

lmoesle committed:
mod: refactor

lmoesle committed:
Merge branch 'epic/error-handling-and-refactoring-integrations-534' into feat/167-create-handle-error-use-case

lmoesle committed:
set release version to 0.18.0

darenegade commented on docs/src/documentation/libs/digiwf-integration-lib/README.md:
```diff
@@ -0,0 +1,169 @@
+
+#### Fachliche Fehler
+
+Fachliche Fehler können über die `TechnicalError`-Exception geworfen werden.
```
Das Naming ist nicht so ideal. Einen fachlichen Fehler "TechnicalError" zu nennen führt sicher zu Verwirrung 😅

darenegade commented on digiwf-integrations/digiwf-example-integration/digiwf-example-integration-example/src/main/java/io/muenchendigital/digiwf/example/integration/streaming/Consumers.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @Bean
+    public Consumer<Message<Object>> onIncidentConsumer() {
```
Etwas komisch, dass ein "TechnicalError" ein eigenes DTO hat und ein Incident nicht 🤔

Soll hier wirklich alles erlaubt sein?

darenegade commented on digiwf-libs/digiwf-integration-lib/digiwf-integration-adapter-springcloudstream/src/main/java/io/muenchendigital/digiwf/integration/adapter/api/SpringCloudStreamAdapter.java:
```diff
@@ -0,0 +1,77 @@
+            try {
+                final String processInstance = processInstanceHeader.orElseThrow(() -> new RuntimeException("Process instance id is missing in message headers. Message will be ignored.")).toString();
+                final String messageName = messageNameHeader.orElseThrow(() -> new RuntimeException("Message type is missing in message headers. Message will be ignored.")).toString();
+                final String messageType = typeHeader.orElseThrow(() -> new RuntimeException("Message type is missing in message headers. Message will be ignored.")).toString();
```
Soll es wirklich immer einen Typ geben🤔

darenegade commented on digiwf-libs/digiwf-integration-lib/digiwf-integration-core/src/main/java/io/muenchendigital/digiwf/integration/core/impl/Integration.java:
```diff
@@ -0,0 +1,40 @@
+     * @throws IllegalAccessException    if this Method object is enforcing Java language access control and the underlying method is inaccessible.
+     */
+    public Object execute(final Object data) throws InvocationTargetException, IllegalAccessException {
+        return this.getMethod().invoke(this.instance, data);
```
Das ist nicht ganz so elegant 😅

Für so einen Anwendungsfall wurde doch SPI erfunden. Gibt es sicher auch moderne Umsetzungen von

darenegade commented:
Bevor ich weiter mache mit dem Review müssten wir mal sprechen.... 😊

lmoesle commented:
Feedback Todos:
- Remove digiwf-integration-lib
- Make sure that digiwf-message supports all functionalities of spring-cloud-stream-utils
- Add Spring Cloud Stream to digiwf-message-core and simplify digiwf-message package structure (core, starter, example)
- Provide Spring Cloud Stream default error handler and make sure retries still work
- Create exception classes for incidentens and technical errors (and rename technical errors)

lmoesle committed:
feat: simplify digiwf-message structure, added error classes, moved spring cloud stream logic to digiwf-message-core

lmoesle committed:
fix: autoconfiguration for simplified application structure

lmoesle committed:
feat: merge examples to one single example and some refactoring

lmoesle committed:
feat: merge examples to one single example and some refactoring

lmoesle committed:
test: added unit tests for MessagePortImpl

lmoesle committed:
Merge branch 'epic/error-handling-and-refactoring-integrations-534' into feat/167-create-handle-error-use-case

lmoesle committed:
Merge branch 'epic/error-handling-and-refactoring-integrations-534' into feat/167-create-handle-error-use-case

# Conflicts:
#	docs/src/guides/custom-integration-service/README.md

lmoesle committed:
feat(digiwf-message): simplify application structure

lmoesle committed:
feat(digiwf-message): simplify application structure

lmoesle committed:
feat(digiwf-message): simplify application structure

lmoesle committed:
fix(digiwf-message): refactor digiwf-message example application

lmoesle committed:
feat: create example integration to demonstrate and test integration concepts

lmoesle committed:
feat: implement error handling with spring cloud stream custom error handler

lmoesle committed:
feat: add process for example integration and refactoring

lmoesle committed:
feat: removed error handler

lmoesle committed:
feat: move error related code to ErrorApi

lmoesle committed:
feat(example-integration): implement error handling

lmoesle committed:
Merge branch 'epic/error-handling-and-refactoring-integrations-534' into feat/167-create-handle-error-use-case

# Conflicts:
#	digiwf-apps/pom.xml

lmoesle committed:
docs: refactoring docs structure for integrations

lmoesle committed:
docs: digiwf-message lib

lmoesle committed:
docs: create integration service guide and integration concept

lmoesle commented:
> 

The custom error handler for spring cloud stream is an alternative to dead letter queuing.
It also has side effects if you combine it with retries. In the case of retries every error (BpmnErrors and IncidentErrors too) are retried before sending them to the custom error handler.
Therefore I dropped the implementation of the error handler and moved back to the simple try catch blocks.

lmoesle committed:
docs: add README.md for digiwf-message lib

lmoesle committed:
Merge branch 'epic/error-handling-and-refactoring-integrations-534' into feat/167-create-handle-error-use-case

lmoesle committed:
docs: add badge

darenegade commented on docs/src/guides/custom-integration-service/README.md:
```diff
@@ -1,3 +1,3 @@
 
-tbd.
\ No newline at end of file
+tbd.
```
Wird das noch benötigt, oder sollte hier dann nicht der Guide stehen, den du unter concept hast?

darenegade reviewed (approved):
Ich finds gut. Vor allem die Doku ist verständlich geschrieben und zusammen mit dem Exaample sollte jeder so Integrationen schnell erstellen können 👍

lmoesle committed:
docs: removed unused code
